packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -1583,14 +1609,9 @@ function commitAttachRef(finishedWork: Fiber) {
       } else {
         retVal = ref(instanceToUse);
       }
-      if (__DEV__) {
-        if (typeof retVal === 'function') {
-          console.error(
-            'Unexpected return value from a callback ref in %s. ' +
-              'A callback ref should not return a function.',
-            getComponentNameFromFiber(finishedWork),
-          );
-        }
+
+      if (typeof retVal === 'function') {
```

Comment:
Normally it's good to check the type as close to where you get it from the user as possible but it'd probably be ok to do this type check during the cleanup instead.

That way there's no extra work on initial mount.

The VM has to check the type anyway before calling it upon cleanup.